### PR TITLE
Add "Code Structure" documentation

### PR DESCRIPTION
**Pull request by pirz**
### What changes were proposed in this pull request?
Adding documentation for 'code structure' document under Hyperspace getting started.

### Why are the changes needed?
To expand Hyperspace documentation and  help new contributors to ramp up.

### Does this PR introduce _any_ user-facing change?
Yes, it changes Hyperspace documentation and extends it.

### How was this patch tested?
Thru proof reading.

### CODE REVIEW

File: docs/_docs/07-dg-code-structure.md [hunk 1]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
 toc: true
```
Comment (apoorvedave1) [inline docs/_docs/07-dg-code-structure.md:6]: update date as well?

**Comment by rapoth** [comment]
Can you fill up the PR description as well when you get a chance?

File: docs/_docs/07-dg-code-structure.md [hunk 2]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
 ---
 
+Hyperspace code is written in Scala and it is using 'sbt' as the build tool. Below is the list of folders in the code base hierarchy:
+### docs
+“docs” folder should be the starting point for looking into Hyperspace code base. It contains useful documentations on different aspects of Hyperspace project, including coding and formatting guidelines, Hyperspace configurations and information on how to build the project in various environments.
```
Comment (rapoth) [inline docs/_docs/07-dg-code-structure.md:11]:
```suggestion
“docs” folder contains useful documentation on different aspects of Hyperspace including coding, contribution and formatting guidelines, and information on how to build the project in various environments. You can access the documentation using the [Hyperspace Documentation](https://microsoft.github.io/hyperspace/docs/ug-quick-start-guide/) site.
```

File: docs/_docs/07-dg-code-structure.md [hunk 3]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+Hyperspace code is written in Scala and it is using 'sbt' as the build tool. Below is the list of folders in the code base hierarchy:
+### docs
+“docs” folder should be the starting point for looking into Hyperspace code base. It contains useful documentations on different aspects of Hyperspace project, including coding and formatting guidelines, Hyperspace configurations and information on how to build the project in various environments. 
+### src
+“src” folder contains Hyperspace’s source and test code. 
+The source code is available under:
+* src/main/scala/com/microsoft/hyperspace
+
+The source code includes implementation of core functionality and features of Hyperspace and is organized in different packages:
+- **actions** : It includes the implementation of Hyperspace actions to create, maintain and remove indexes in a concurrent manner.
```
Comment (rapoth) [inline docs/_docs/07-dg-code-structure.md:18]:
```suggestion
- **actions** : Implementation of Hyperspace actions to create, maintain and remove indexes in a concurrent manner.
```

File: docs/_docs/07-dg-code-structure.md [hunk 4]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+- **index**: It includes implementation of Hyeprspace indexes, optimizations rules and other core features.
```
Comment (rapoth) [inline docs/_docs/07-dg-code-structure.md:19]:
```suggestion
- **index**: Implementation of Hyeprspace indexes, optimizations rules and other core features. 
```

File: docs/_docs/07-dg-code-structure.md [hunk 5]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+“docs” folder should be the starting point for looking into Hyperspace code base. It contains useful documentations on different aspects of Hyperspace project, including coding and formatting guidelines, Hyperspace configurations and information on how to build the project in various environments. 
+### src
+“src” folder contains Hyperspace’s source and test code. 
+The source code is available under:
+* src/main/scala/com/microsoft/hyperspace
+
+The source code includes implementation of core functionality and features of Hyperspace and is organized in different packages:
+- **actions** : It includes the implementation of Hyperspace actions to create, maintain and remove indexes in a concurrent manner.
+- **index**: It includes implementation of Hyeprspace indexes, optimizations rules and other core features. 
+- **util**: Various utility codes that are used by different features and components are added here.
```
Comment (rapoth) [inline docs/_docs/07-dg-code-structure.md:20]:
```suggestion
- **util**: Various utility functions that are used by different features and components.
```

**Comment by imback82** [comment]
Can you fill out the description please?

File: docs/_docs/07-dg-code-structure.md [hunk 6]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+
+If you are working on adding changes to Hyeprspace, you would need to add or modify code in above packages.
```
Comment (rapoth) [inline docs/_docs/07-dg-code-structure.md:22]:
```suggestion
If you are working on adding changes to Hyperspace, you would need to add or modify code in above packages.
```

File: docs/_docs/07-dg-code-structure.md [hunk 7]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+
+The test code is available under:
+* src/test/scala/com/microsoft/hyperspace
+
+The test code has a similar structure as the source code and each package in the test code includes unit test cases for the corresponding feature or functionality from the source code. Any code change or additional feature implementation needs to add relevant cases which fully cover the modified code.
+### dev
+“dev” folder contains required resources for contributing code to the Hyperspace project as a developer. Currently, it includes the Scala formatting configuration file that is needed to make sure any new code change complies with the Hyperspace’s coding guidelines.
```
Comment (rapoth) [inline docs/_docs/07-dg-code-structure.md:29]:
```suggestion
“dev” folder contains required resources for contributing code to the Hyperspace. Currently, it includes the Scala formatting configuration file that is needed to make sure any new code change complies with Hyperspace’s [coding guidelines](https://microsoft.github.io/hyperspace/docs/dg-contributing/).
```

File: docs/_docs/07-dg-code-structure.md [hunk 8]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+### build
+Hyperspace uses “sbt” is the build tool for building its code. “build” folder contains sbt build scripts and configurations which are used to create an artifact from Hyperspace code.
```
Comment (rapoth) [inline docs/_docs/07-dg-code-structure.md:31]:
```suggestion
Hyperspace uses “sbt” as the build tool for building its code. “build” folder contains sbt build scripts and configurations which are used to create an artifact from Hyperspace source code.
```

File: docs/_docs/07-dg-code-structure.md [hunk 9]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
 toc: true
 ---
 
+Hyperspace code is written in Scala and it is using 'sbt' as the build tool. Below is the list of folders in the code base hierarchy:
+### docs
+“docs” folder should be the starting point for looking into Hyperspace code base. It contains useful documentations on different aspects of Hyperspace project, including coding and formatting guidelines, Hyperspace configurations and information on how to build the project in various environments. 
+### src
```
Comment (imback82) [inline docs/_docs/07-dg-code-structure.md:12]: Can we move `src` to the top?
Also can you add some new lines around ### to make it more readable?

File: docs/_docs/07-dg-code-structure.md [hunk 10]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+“src” folder contains Hyperspace’s source and test code. 
+The source code is available under:
+* src/main/scala/com/microsoft/hyperspace
+
+The source code includes implementation of core functionality and features of Hyperspace and is organized in different packages:
+- **actions** : It includes the implementation of Hyperspace actions to create, maintain and remove indexes in a concurrent manner.
+- **index**: It includes implementation of Hyeprspace indexes, optimizations rules and other core features. 
+- **util**: Various utility codes that are used by different features and components are added here.
+
```
Comment (imback82) [inline docs/_docs/07-dg-code-structure.md:21]: how about code at `src/main/scala/com/microsoft/hyperspace`? and `org/apache/spark`?

File: docs/_docs/07-dg-code-structure.md [hunk 11]
```diff
@@ -6,3 +6,28 @@ last_modified_at: 2020-06-20
...
+The test code is available under:
+* src/test/scala/com/microsoft/hyperspace
+
+The test code has a similar structure as the source code and each package in the test code includes unit test cases for the corresponding feature or functionality from the source code. Any code change or additional feature implementation needs to add relevant cases which fully cover the modified code.
+### dev
+“dev” folder contains required resources for contributing code to the Hyperspace project as a developer. Currently, it includes the Scala formatting configuration file that is needed to make sure any new code change complies with the Hyperspace’s coding guidelines.
+### build
+Hyperspace uses “sbt” is the build tool for building its code. “build” folder contains sbt build scripts and configurations which are used to create an artifact from Hyperspace code.
+### project
+“project” folder contains additional sbt build properties and plugin configurations which are needed to correctly format Hyperspace code and build it to generate a Hyperspace artifact.
```
Comment (imback82) [inline docs/_docs/07-dg-code-structure.md:33]: `“project” folder contains additional sbt build properties and plugin configurations.` is good enough.

File: docs/_docs/07-dg-code-structure.md (hunk 1 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md:6]: Done

**Comment by pirz** [comment]
> Can you fill up the PR description as well when you get a chance?

@rapoth  Done

File: docs/_docs/07-dg-code-structure.md (hunk 2 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 3 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 4 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 5 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 6 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 7 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 8 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 10 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md:21]: Done

File: docs/_docs/07-dg-code-structure.md (hunk 9 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md:11]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 11 above)
Comment (pirz) [reply docs/_docs/07-dg-code-structure.md]: Done.

File: docs/_docs/07-dg-code-structure.md (hunk 10 above)
Comment (imback82) [reply docs/_docs/07-dg-code-structure.md:21]: How about something like:

**`src/main/scala/com/microsoft/hyperspace`**
&nbsp;&nbsp;&nbsp;|- **`actions`**: blah
&nbsp;&nbsp;&nbsp;|- **`index`**: blah
**`src/main/scala/org/apache/spark`**
&nbsp;&nbsp;&nbsp;|- **`dir`**: blah

Comment (pirz) [reply docs/_docs/07-dg-code-structure.md:24]: Done.

**Pull request merged**